server/api.go

```diff
@@ -0,0 +1,1024 @@
+		dialogError(w, err.Error(), nil)
+	}
+
+	p.mm.Post.SendEphemeralPost(userID, &model.Post{
+		UserId:    p.BotUserID,
+		ChannelId: req.ChannelId,
+		Message:   "Subscription added",
+	})
+
+	dialogOK(w)
+}
+
+func (p *Plugin) dialogDeleteSubscription(w http.ResponseWriter, r *http.Request, userID string) {
+	req := model.SubmitDialogRequestFromJson(r.Body)
+	if req == nil {
+		dialogError(w, "could not get the dialog request", nil)
+		return
+	}
+
+	u, err := p.mm.User.Get(userID)
+	if err != nil {
+		dialogError(w, err.Error(), nil)
+		return
+	}
+
+	if !canCreateSubscription(*u, req.ChannelId) {
+		dialogError(w, "You cannot delete a subscription", nil)
+		return
+	}
+
+	typeIDStr, errText, errors := getDialogSubmissionTextField(req, DialogFieldBadgeType)
+	if errors != nil {
+		dialogError(w, errText, errors)
+		return
+	}
+
+	typeID, err := strconv.Atoi(typeIDStr)
+	if err != nil {
+		dialogError(w, "Invalid field", map[string]string{DialogFieldBadgeType: "Invalid type"})
+		return
+	}
+
+	err = p.store.RemoveSubscriptions(badgesmodel.BadgeType(typeID), req.ChannelId)
+	if err != nil {
+		dialogError(w, err.Error(), nil)
+	}
+
+	p.mm.Post.SendEphemeralPost(userID, &model.Post{
+		UserId:    p.BotUserID,
+		ChannelId: req.ChannelId,
+		Message:   "Subscription removed",
+	})
+
+	dialogOK(w)
+}
+
+func getDialogSubmissionTextField(req *model.SubmitDialogRequest, fieldName string) (value string, errText string, errors map[string]string) {
+	value, ok := req.Submission[fieldName].(string)
+	value = strings.TrimSpace(value)
+	if !ok || value == "" {
+		return "", "Invalid argument", map[string]string{fieldName: "Field empty or not recognized."}
+	}
+
+	return value, "", nil
+}
+
+func getDialogSubmissionBoolField(req *model.SubmitDialogRequest, fieldName string) bool {
+	value, _ := req.Submission[fieldName].(bool)
+	return value
+}
+
+func (p *Plugin) grantBadge(w http.ResponseWriter, r *http.Request, pluginID string) {
+	var req *badgesmodel.GrantBadgeRequest
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		p.writeAPIError(w, &APIErrorResponse{
+			ID:         "cannot unmarshal request",
+			Message:    err.Error(),
+			StatusCode: http.StatusInternalServerError,
+		})
+		return
+	}
+	p.mm.Log.Debug("Granting badge", "req", req)
+
+	if req == nil {
+		p.writeAPIError(w, &APIErrorResponse{
+			ID:         "missing request",
+			Message:    "Missing grant request on request body",
+			StatusCode: http.StatusInternalServerError,
+		})
+		return
+	}
+
+	shouldNotify, err := p.store.GrantBadge(req.BadgeID, req.UserID, req.BotID, req.Reason)
+	if err != nil {
+		p.writeAPIError(w, &APIErrorResponse{
+			ID:         "cannot grant badge",
+			Message:    err.Error(),
+			StatusCode: http.StatusInternalServerError,
+		})
+		return
+	}
+	if shouldNotify {
+		u, err := p.mm.User.Get(req.UserID)
+		if err == nil {
+			p.notifyGrant(req.BadgeID, req.BotID, u, false, "", req.Reason)
+		}
+	}
+
+	_, _ = w.Write([]byte("OK"))
```

Comment:
Nip: If this is a json API it should return JSON, maybe `"{\"success\": true}"` or something like that.